src/coreclr/vm/frozenobjectheap.cpp

```diff
@@ -135,34 +157,65 @@ FrozenObjectSegment::FrozenObjectSegment(size_t sizeHint) :
         ThrowOutOfMemory();
     }
 
+    m_pStart = static_cast<uint8_t*>(committedAlloc);
+    m_pCurrent = m_pStart + sizeof(ObjHeader);
+    m_SizeCommitted = FOH_COMMIT_SIZE;
+
     // ClrVirtualAlloc is expected to be PageSize-aligned so we can expect
     // DATA_ALIGNMENT alignment as well
     _ASSERT(IS_ALIGNED(committedAlloc, DATA_ALIGNMENT));
+}
 
-    segment_info si;
-    si.pvMem = committedAlloc;
-    si.ibFirstObject = sizeof(ObjHeader);
-    si.ibAllocated = si.ibFirstObject;
-    si.ibCommit = FOH_COMMIT_SIZE;
-    si.ibReserved = m_Size;
+void FrozenObjectSegment::RegisterOrUpdate(uint8_t* current, size_t sizeCommited)
+{
+    CONTRACTL
+    {
+        THROWS;
+        MODE_PREEMPTIVE;
+    }
+    CONTRACTL_END
 
-    m_SegmentHandle = GCHeapUtilities::GetGCHeap()->RegisterFrozenSegment(&si);
-    if (m_SegmentHandle == nullptr)
+    if (!IsRegistered())
     {
-        ClrVirtualFree(alloc, 0, MEM_RELEASE);
-        ThrowOutOfMemory();
+        // Other threads won't touch these fields until we set m_IsRegistered to true
```

Comment:
This works, but it is not straightforward to see that it is correct when `RegisterFrozenSegment` fails.

It would be easier to see that this is correct if you set these fields after `RegisterFrozenSegment` succeeds. You would not actually even need `m_IsRegistered` bool flag if you do it that way - non-null `m_pCurrentRegistered can act as `m_IsRegistered` flag.